### DISCUSSION TITLE
How to switch userAgent in the middle of tests

### DISCUSSION THREAD

**Discussion by RezaRahmati** [cypress-io/cypress#17558] [Questions and Help]
Hi 

I have below test scenario
open website on desktop > fill information > get the link > open link on mobile > continue the flow

So I need to switch between desktop and mobile in the middle of tests

Based on docs [here](https://docs.cypress.io/api/cypress-api/config#Notes) , userAgent changes are ignore in the run time

I have tried many different like below to change userAgent none is working, and userAgent is read from cypress.json

```

describe('Start From Mobile', () => {
	before(() => {
		Cypress.config('userAgent', 'Mozilla/5.0 (Linux; U; Android 9.0.2; en-us; Galaxy Nexus Build/ICL53F) AppleWebKit/534.30 (KHTML, like Gecko) Version/4.0 Mobile Safari/534.30'); // set userAgent
	});

	it('should get error message to continue on phone', () => {
		cy.viewport('iphone-6')
		cy.log('navigator.userAgent', window.navigator.userAgent);
		cy.visit(Cypress.env('url'), {
			onBeforeLoad: win => {
				Object.defineProperty(win.navigator, 'userAgent', {
					value: window.navigator.userAgent,
				});
			},
		});
		// cy.visit(Cypress.env('url'), {
		// 	headers: {
		// 		'user-agent': window.navigator.userAgent,
		// 	}
		// });
```

Wondering what is the solution to be able to write my tests?

Thanks

**Comment by ksagnard** [comment]
Good afternoon,

I have a similar need to test the following scenario : 

- Open browser
- Check a text is displayed
- Close the browser
- Reopen browser as google bot
- Check another text is displayed

It seems this was possible using the parameter user-agent in Cypress 3.3 but it does no longer work https://stackoverflow.com/questions/51048880/is-there-a-programmatic-way-to-change-user-agent-in-cypress-io 

Do you have a way to do this test?

**Comment by mirobo** [comment]
You should probably follow this issue: https://github.com/cypress-io/cypress/issues/2100

Forget this:
```
cy.visit(Cypress.env('url'), {
			onBeforeLoad: win => {
				Object.defineProperty(win.navigator, 'userAgent', {
					value: window.navigator.userAgent,
				});
			},
		});
```		
because I just tested this in Cypress 10.3.1 and whenever in your test you execute cy.reload() or cy.visit('...') (without onBeforeLoad), the userAgent will jump back to the default in cypress.config.js (or cypress.json for Cypress < 10).

The only way I saw so far is to start Cypress explicitly with another config, i.e.
npx cypress open --browser firefox --config userAgent="Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:91.0) Gecko/20100101 Firefox/91.0"

And then follow instructions here https://docs.cypress.io/guides/guides/cross-browser-testing#Running-Specific-Tests-by-Browser and https://docs.cypress.io/guides/core-concepts/writing-and-organizing-tests#Test-Configuration on how to configure tests to only run with a specific browser. And maybe just seperate mobile tests from desktop tests by folder and set different "specPattern" (or "testFiles" in Cypress < 10) in your cypress.config.js.


Not sure if there are other ways.. maybe @lmiller1990 has some other ideas? :-D

I'll bring in some other ideas into issue #2100
